### ISSUE TITLE
valAcc disparity between flow_from_directory() and flow_from_dataframe()

### ISSUE THREAD

**Issue by Mahi-Mai** [keras-team/keras-preprocessing#102]
In order to test how my model training script performed on a benchmark dataset, I converted the stored MNIST to a set of png images.  I have them organized in two ways:

**Method 1.**  I have a "train" folder and a "test" folder where images are stored without further organization.  I have a dataframe for the train and test set, with column 1 listing the absolute directory, and column 2 listing the label.  I've carefully examined this csv- the labels and image listings appear accurate.  I've captured and tested the output of flow_from_dataframe(), and it looks fine.

Sample of csv:
```
,test_samples,test_labels
0,/path/Data/test/6992.png,7
1,/path/Data/test/1380.png,9
2,/path/Data/test/5817.png,4
3,/path/Data/test/5295.png,5
4,/path/Data/test/5340.png,2
```

**Method 2.**  I have a train and test folder, each with subdirectories for the different categories of images.

Other than how they're organized, these datasets are otherwise identical.

If I run my script and use flow_from_dataframe() with the assets from Method 1, the highest validation accuracy I can manage ranges from 0.01-0.05.  If I run my script with flow_from_directory() using the assets in Method 2, my highest checkpoint is 0.93.

What could be the source of this disparity?  Am I misusing flow_from_dataframe()?  I'll share my scripts from each approach below.  Thanks in advance for any insight.

**Method 1: Garbage Validation Accuracy**
```
import pandas as pd
import numpy as np
 
import keras
from keras_preprocessing.image import ImageDataGenerator
 
from keras import applications
from keras import optimizers
from keras.models import Model 
from keras.layers import Dropout, Flatten, Dense, GlobalAveragePooling2D
from keras import backend as k 
from keras.callbacks import ModelCheckpoint, CSVLogger
 
from keras.applications.vgg16 import VGG16, preprocess_input
 
# INITIALIZE MODEL
 
img_width, img_height = 32, 32
model = VGG16(weights = 'imagenet', include_top=False, input_shape = (img_width, img_height, 3))
 
# freeze all layers
for layer in model.layers:
    layer.trainable = False
 
# Adding custom Layers 
x = model.output
x = Flatten()(x)
x = Dense(512, activation='relu')(x)
x = Dropout(0.5)(x)
predictions = Dense(10, activation="softmax")(x)
 
# creating the final model 
model_final = Model(input = model.input, output = predictions)
 
# compile the model 
rms = optimizers.RMSprop(lr=1e-4)
 
model_final.compile(loss = "categorical_crossentropy", optimizer = rms, metrics=["accuracy"])
 
# LOAD AND DEFINE SOURCE DATA
#df.column_name = df.column_name.astype(str)
 
train = pd.read_csv('/path/Data/MNIST_train.csv', index_col=0)
train.train_labels = train.train_labels.astype(str)
 
val = pd.read_csv('/path/Data/MNIST_test.csv', index_col=0)
val.test_labels = val.test_labels.astype(str)
 
nb_train_samples = 60000
nb_validation_samples = 10000
batch_size = 60
epochs = 5
 
# Initiate the train and test generators
train_datagen = ImageDataGenerator()
test_datagen = ImageDataGenerator()
 
train_generator = train_datagen.flow_from_dataframe(dataframe=train,
                                                    directory=None,
                                                    x_col='train_samples',
                                                    y_col='train_labels',
                                                    has_ext=True,
                                                    target_size = (img_height,
                                                                   img_width),
                                                    batch_size = batch_size, 
                                                    class_mode = 'categorical',
                                                    color_mode = 'rgb')
 
validation_generator = test_datagen.flow_from_dataframe(dataframe=val,
                                                        directory=None,
                                                        x_col='test_samples',
                                                        y_col='test_labels',
                                                        has_ext=True,
                                                        target_size = (img_height, 
                                                                       img_width),
                                                        batch_size = batch_size, 
                                                        class_mode = 'categorical',
                                                        color_mode = 'rgb')

# DEFINE CALLBACKS
 
path = '/path/chk/epoch_{epoch:02d}-valLoss_{val_loss:.2f}-valAcc_{val_acc:.2f}.hdf5'
 chk = ModelCheckpoint(path, monitor = 'val_acc', verbose = 1, save_best_only = True, mode = 'max')
 logger = CSVLogger('/path/chk/training_log.csv', separator = ',', append=False)
 
nPlus = 1
samples_per_epoch = nb_train_samples * nPlus
 
# Train the model 
model_final.fit_generator(train_generator,
                          steps_per_epoch = int(samples_per_epoch/batch_size),
                          epochs = epochs,
                          validation_data = validation_generator,
                          validation_steps = int(nb_validation_samples/batch_size),
                          callbacks = [chk, logger])
```

**METHOD 2: Stellar Validation Accuracy**
```
from keras import applications
from keras.preprocessing.image import ImageDataGenerator
from keras import optimizers
from keras.models import Model 
from keras.layers import Dropout, Flatten, Dense, GlobalAveragePooling2D
from keras import backend as k 
from keras.callbacks import ModelCheckpoint, CSVLogger
 
img_width, img_height = 32, 32
train_data_dir = '/path/Data/categorical_subdirectories/test/'
validation_data_dir = '/path/Data/categorical_subdirectories/train/'
nb_train_samples = 60000
nb_validation_samples = 10000
batch_size = 60
epochs = 10
 
from keras.applications.vgg16 import VGG16, preprocess_input
 
model = VGG16(weights = "imagenet", include_top=False, input_shape = (img_width, img_height, 3))
 
 # Freeze the layers which you don't want to train. Here I am freezing the first 5 layers.
for layer in model.layers:
    layer.trainable = False
 
#Adding custom Layers 
x = model.output
x = Flatten()(x)
x = Dense(512, activation='relu')(x)
x = Dropout(0.5)(x)
predictions = Dense(10, activation="softmax")(x)
 
# creating the final model 
model_final = Model(input = model.input, output = predictions)
 
RMSprop = optimizers.RMSprop(lr=1e-4)
 
# compile the model 
model_final.compile(loss = "categorical_crossentropy", optimizer = RMSprop, metrics=["accuracy"])
 
model_final.summary()
 
# Initiate the train and test generators with data Augumentation 
train_datagen = ImageDataGenerator(preprocessing_function = preprocess_input)
test_datagen = ImageDataGenerator(preprocessing_function = preprocess_input)
 
train_generator = train_datagen.flow_from_directory(train_data_dir,
                                                    target_size = (img_height, img_width),
                                                    batch_size = batch_size, 
                                                    class_mode = "categorical")
 
validation_generator = test_datagen.flow_from_directory(validation_data_dir,
                                                        target_size = (img_height, img_width),
                                                        class_mode = "categorical")
 
# Save the model according to the conditions
path = '/path/chk/epoch_{epoch:02d}-valLoss_{val_loss:.2f}-valAcc_{val_acc:.2f}.hdf5'
chk = ModelCheckpoint(path, monitor = 'val_acc', verbose = 1, save_best_only = True, mode = 'max')
logger = CSVLogger('/path/chk/training log.csv', separator = ',', append=False)
 
nPlus = 1
samples_per_epoch = nb_train_samples * nPlus
 
# Train the model 
model_final.fit_generator(train_generator,
                          steps_per_epoch = int(samples_per_epoch/batch_size),
                          epochs = epochs,
                          validation_data = validation_generator,
                          validation_steps = int(nb_validation_samples/batch_size),
                          callbacks = [chk, logger])
```

**Comment by smurak** [comment]
I guess the reason is an issue mentioned in #98.
A workaround is set the `classes` option explicitly.

**Comment by Mahi-Mai** [comment]
I can confirm that flow_from_dataframe() works when the classes are explicitly defined for the train and validation generator.  Woo!!

**Comment by Mahi-Mai** [comment]
@smurak   What if I use flow_from_dataframe() in a regression problem?  Where:
```
...
predictions = Dense(1, activation='linear')(x)
model_final = Model(input = model.input, output = predictions)
model_final.compile(loss = 'mse', optimizer = RMSprop, metrics=['mse', 'mae'])
```

and the validation generator is defined like this:
```
validation_generator = val_datagen.flow_from_dataframe(dataframe=val,
                                                       directory=None,
                                                       x_col='directory',
                                                       y_col='score',
                                                       has_ext=True,
                                                       target_size = (img_height, img_width),
                                                       batch_size = batch_size,
                                                       class_mode = 'other',
                                                       shuffle=False)
```

How then do I validate that the right labels are being pulled for my samples?

**Comment by smurak** [comment]
@Mahi-Mai  I haven't used it for regression problems, but I guess just put the generator to `fit_generator` will work.
If you have some problems, please write the output of `next(validation_generator)` and more details about the problems.

**Comment by rragundez** [comment]
This issuea has been resolved and should be closed @Dref360
